### ISSUE TITLE
kernel BUG at /root/ab/tempesta-ab/tempesta_fw/http.c:1321

### ISSUE THREAD

**Issue by keshonok** [tempesta-tech/tempesta#696]
The setup:
* Tempesta's current master at c40924b;
* Tempesta kernel 4.8.15 at [5b3aa7f](https://github.com/tempesta-tech/linux-4.8.15-tfw/commit/5b3aa7fe30ce46bb8f112a9544548da926cb3714);
* Both Tempesta and Nginx run on the same physical test server with 12 CPU cores and 64Gb RAM;
* Simplest Tempesta configuration:
```
server 127.0.0.1:8080
```

The configuration means that Tempesta bypasses cache on all requests. All requests are forwarded to the back-end server without exceptions.

The test is a simple `wrk -t 8 -c 1024 -d 600s http://192.168.0.1`. It runs on another physical machine with 8 CPU cores and 32Gb RAM. It is connected with the test server via 10G interface.

```
[  632.352381] kernel BUG at /root/ab/tempesta-ab/tempesta_fw/http.c:1321!
[  632.359233] invalid opcode: 0000 [#1] SMP
[  632.363473] Modules linked in: tfw_sched_rr(O) tfw_sched_http(O) tfw_sched_hash(O) tempesta_fw(O) tempesta_db(O) tempesta_tls(O) joydev hid_generic usbhid hid mlx4_en iTCO_wdt iTCO_vendor_support intel_rapl evdev ehci_pci sb_edac edac_core x86_pkg_temp_thermal ast intel_powerclamp igb coretemp ehci_hcd ttm snd_pcm crct10dif_pclmul crc32_pclmul crc32c_intel ghash_clmulni_intel snd_timer aesni_intel drm_kms_helper ptp snd aes_x86_64 glue_helper lrw gf128mul ablk_helper xhci_pci xhci_hcd drm ahci cryptd pps_core mlx4_core soundcore libahci usbcore pcspkr i2c_algo_bit libata fb_sys_fops sg i2c_i801 syscopyarea sysfillrect i2c_smbus sysimgblt mei_me ioatdma i2c_core mei lpc_ich dca usb_common mfd_core wmi ipmi_si ipmi_msghandler acpi_power_meter acpi_pad tpm_tis tpm_tis_core tpm button autofs4 ext4 crc16 jbd2 mbcache sd_mod mpt3sas raid_class scsi_transport_sas scsi_mod shpchp
[  632.444601] CPU: 7 PID: 52 Comm: ksoftirqd/7 Tainted: G           O    4.8.15-debug-ab+ #7
[  632.453339] Hardware name: Supermicro Super Server/X10SRH-CF, BIOS 2.0 12/17/2015
[  632.461315] task: ffff881017839780 task.stack: ffff881017848000
[  632.467743] RIP: 0010:[<ffffffffa0a90b21>]  [<ffffffffa0a90b21>] tfw_http_conn_drop+0x151/0x220 [tempesta_fw]
[  632.478231] RSP: 0018:ffff88101784b710  EFLAGS: 00010287
[  632.484101] RAX: ffff880f2ec4b020 RBX: ffff880f3487a490 RCX: ffff880f2ec4b3e0
[  632.491785] RDX: ffff880f2ea223e0 RSI: 0000000000000001 RDI: ffff880f3487a560
[  632.499484] RBP: ffff88101784b760 R08: 0000000000000001 R09: 0000000000000000
[  632.507192] R10: 0000000000000000 R11: 0000000000000000 R12: ffff880f3487a550
[  632.514910] R13: ffff88101784b710 R14: ffff880f3487a560 R15: ffff880f47830480
[  632.522645] FS:  0000000000000000(0000) GS:ffff88103f3c0000(0000) knlGS:0000000000000000
[  632.531355] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  632.537734] CR2: 00007fa37942f5d8 CR3: 000000100260f000 CR4: 00000000001406e0
[  632.545521] Stack:
[  632.548222]  ffff880f2ec4b020 ffff880f2ec4b020 0000000000000296 000000004c4adbe6
[  632.556425]  ffff881017839780 ffff880f3487a490 ffff8810033b4010 ffff880f3487a4c8
[  632.564617]  0000000000000000 ffff880f47830480 ffff88101784b780 ffffffffa0a8cf9a
[  632.572805] Call Trace:
[  632.575962]  [<ffffffffa0a8cf9a>] tfw_connection_drop+0x2a/0x40 [tempesta_fw]
[  632.583827]  [<ffffffff815fa417>] ? _raw_spin_unlock+0x27/0x40
[  632.590402]  [<ffffffffa0aab045>] tfw_sock_clnt_do_drop.isra.8+0x75/0xb0 [tempesta_fw]
[  632.599085]  [<ffffffffa0aab08e>] tfw_sock_clnt_error+0xe/0x10 [tempesta_fw]
[  632.606912]  [<ffffffffa0aa9713>] ss_linkerror+0x33/0x70 [tempesta_fw]
[  632.614229]  [<ffffffffa0aaa0a8>] ss_tcp_state_change+0xa8/0x2d0 [tempesta_fw]
[  632.622251]  [<ffffffff81531d86>] tcp_fin+0x146/0x3c0
[  632.628126]  [<ffffffff8153278c>] tcp_data_queue+0x78c/0xf00
[  632.634603]  [<ffffffff814ef2ba>] ? sk_filter_trim_cap+0x4a/0x390
[  632.641508]  [<ffffffff81533132>] tcp_rcv_established+0x232/0x730
[  632.648446]  [<ffffffff8153eb31>] tcp_v4_do_rcv+0x141/0x190
[  632.654840]  [<ffffffff8153ffd4>] tcp_v4_rcv+0xa64/0xca0
[  632.660997]  [<ffffffff8151617f>] ? ip_local_deliver_finish+0x2f/0x380
[  632.668405]  [<ffffffff81516225>] ip_local_deliver_finish+0xd5/0x380
[  632.675604]  [<ffffffff8151617f>] ? ip_local_deliver_finish+0x2f/0x380
[  632.682994]  [<ffffffff8151671f>] ip_local_deliver+0x6f/0xf0
[  632.689528]  [<ffffffff8153f4ea>] ? tcp_v4_early_demux+0xfa/0x180
[  632.696537]  [<ffffffff81515cb7>] ip_rcv_finish+0x197/0x630
[  632.703002]  [<ffffffff81516a2b>] ip_rcv+0x28b/0x3d0
[  632.708905]  [<ffffffff81515b20>] ? inet_del_offload+0x40/0x40
[  632.715634]  [<ffffffff814d2b1a>] __netif_receive_skb_core+0x31a/0x9c0
[  632.723051]  [<ffffffff814d4800>] ? netif_receive_skb_internal+0xe0/0x200
[  632.730723]  [<ffffffff814cc878>] ? get_rps_cpu+0x78/0x480
[  632.737087]  [<ffffffff814d31d8>] __netif_receive_skb+0x18/0x60
[  632.743851]  [<ffffffff814d4792>] netif_receive_skb_internal+0x72/0x200
[  632.751286]  [<ffffffff814d4745>] ? netif_receive_skb_internal+0x25/0x200
[  632.758875]  [<ffffffff814d5d87>] napi_gro_frags+0xf7/0x2c0
[  632.765222]  [<ffffffffa04207c6>] mlx4_en_process_rx_cq+0x496/0xf20 [mlx4_en]
[  632.773133]  [<ffffffffa04212e0>] mlx4_en_poll_rx_cq+0x30/0x90 [mlx4_en]
[  632.780608]  [<ffffffff814d549f>] net_rx_action+0x22f/0x460
[  632.786904]  [<ffffffff810e7924>] ? rcu_process_callbacks+0x6b4/0x800
[  632.794056]  [<ffffffff815fd39d>] __do_softirq+0xcd/0x4ad
[  632.800173]  [<ffffffff81096734>] ? smpboot_thread_fn+0x34/0x1e0
[  632.806868]  [<ffffffff81096829>] ? smpboot_thread_fn+0x129/0x1e0
[  632.813634]  [<ffffffff8107393f>] run_ksoftirqd+0x1f/0x60
[  632.819747]  [<ffffffff81096824>] smpboot_thread_fn+0x124/0x1e0
[  632.826331]  [<ffffffff81096700>] ? sort_range+0x30/0x30
[  632.832329]  [<ffffffff81092c2e>] kthread+0xfe/0x120
[  632.837946]  [<ffffffff815fad6f>] ret_from_fork+0x1f/0x40
[  632.843993]  [<ffffffff81092b30>] ? kthread_create_on_node+0x240/0x240
[  632.851164] Code: 08 eb d1 4c 89 f7 e8 ef 98 b6 e0 48 8b 45 b0 4c 39 e8 0f 84 05 ff ff ff 48 8b 90 c0 03 00 00 48 8d 88 c0 03 00 00 48 39 ca 74 14 <0f> 0b 4c 3b a7 c8 00 00 00 0f 85 62 ff ff ff e9 de fe ff ff 48 
[  632.872455] RIP  [<ffffffffa0a90b21>] tfw_http_conn_drop+0x151/0x220 [tempesta_fw]
[  632.880752]  RSP <ffff88101784b710>
```

**Comment by keshonok** [comment]
The bug is caused by a combination of two major factors:
* `TfwCliConn{}`  has a union for `resp` and `reason`.
*  When a client connection is dropped, `req->resp` is used to distinguish connections that are NOT in the forwarding queue.

The assumption was that `resp` and `reason` are never used both at the same time. That's true indeed. When a client connection is dropped some requests may be "in flight" - about to be dropped with an HTTP error response sent to the client. In that case `req->resp` is actually `req->reason`. 

`req->resp` is updated under `req->msg.seq_lock`, and the same lock is used in `tfw_http_conn_cli_drop()`. That is intended to ensure that when `req->resp` is set, then the request is definitely NOT in a forwarding queue.

However, `tfw_http_conn_cli_drop()` has no ability to know that `req->resp` may actually be `req->reason`. Moreover, `req->reason` is usually updated under a completely different lock.

`req->resp` and `req-reason` may not be shared. A proper fix is in the works.

**Comment by keshonok** [comment]
There are multiple possible implementations that solve this issue. The simplest is to decouple `req->reason` from `req->resp` and make them separate members of `TfwHttpReq{}`. Also, decouple `req->status` and `req->retries` at the same time. However, there may be hundreds of thousand of connections, and it's essential that the size of `TfwHttpReq{}` is kept to a necessary minumum. The structure shouldn't contain short-lived temporary data like `status` and `reason`.

These temporary data need to be kept somewhere else and linked with an appropriate `TfwHttpReq{}` instance (`req` pointer). The data can be put in an allocated memory, and Tempesta has a fast way to allocate memory from memory pool that every request instance has.  The major issue here is how to deal with memory allocation error, even if that situation would be quite rare.

If memory allocation fails, then Tempesta fails to save data for an HTTP error response, and therefore the HTTP error response will not be sent at all for a corresponding request. Then we have the same issue as the one described in #699. However, to deal with this error we can't just call `tfw_http_resp_fwd()` with NULL `resp` pointer as suggested in #699. The code that saves the data for an HTTP error response is usually run under a server connection's queue lock `fwd_qlock`, while `tfw_http_resp_fwd()` takes a client connection's queue lock `seq_qlock`. These locks should not be chained, and the rule is that they are not taken both at the same time.

In the end, the most reasonable solution is to leave error handling to the actual sending of an HTTP error response as described in #699, and avoid any chance for an error when saving data for an error response. Don't allocate extra memory for the saved data, and just use an unused area in `TfwHttpReq{}`. That area is `TfwHttpReq{}->parser` which keeps the parser state while an HTTP message is parsed. At the time the data for an HTTP error response is saved the message had been completely parsed, and so that space may be reused.

**Comment by keshonok** [comment]
Fixed in 001eb12.
